tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -475,6 +477,18 @@ void hlsl::DiagnoseTranslationUnit(clang::Sema *self) {
       HLSLMethodCallDiagnoseVisitor Visitor(self, shaderModel, EntrySK, FDecl,
                                             DiagnosedCalls);
       Visitor.TraverseDecl(FD);
+
+      // diagnose any node functions that have incompatible launch types with
+      // the present SV semantics on each parameter.
+      if (EntrySK == DXIL::ShaderKind::Node) {
+        if (const auto *NodeLaunchAttr =
+                FDecl->getAttr<clang::HLSLNodeLaunchAttr>()) {
+          llvm::StringRef NodeLaunchTyStr = NodeLaunchAttr->getLaunchType();
+          hlsl::DXIL::NodeLaunchType NodeLaunchTy =
```

Comment:
Is it possible to do this in Sema instead of hlsl::DiagnoseTranslationUnit?

Reply:
I don't think we can do this in sema because we have to account for library diagnostics. Sema has DiagnoseEntry, which runs on shaders that are not the library shader kind. SemaHLSLDiagnoseTU runs with library shader kinds. So we have to add this here to account for the case where a shader in a library file has the node shader attribute and an invalid nodeLaunch type.